sling/nlp/kb/phrase-table-builder.cc

```diff
@@ -0,0 +1,466 @@
+#include "sling/task/task.h"
+#include "sling/util/mutex.h"
+
+namespace sling {
+namespace nlp {
+
+// Build phrase table repository from aliases.
+class PhraseTableBuilder : public task::FrameProcessor {
+ public:
+  void Startup(task::Task *task) override {
+    // Get language for names.
+    string lang = task->Get("language", "en");
+    language_ = commons_->Lookup("/lang/" + lang);
+    task->Fetch("reliable_alias_sources", &reliable_alias_sources_);
+    task->Fetch("transfer_aliases", &transfer_aliases_);
+
+    // Set phrase normalization.
+    tokenizer_.set_normalization(
+        ParseNormalization(task->Get("normalization", "lcn")));
+
+    // Initialize alias transfer.
+    if (transfer_aliases_) InitAliasTransfer();
+
+    // Statistics.
+    num_aliases_ = task->GetCounter("aliases");
+    num_phrases_ = task->GetCounter("phrases");
+    num_entities_ = task->GetCounter("entities");
+    num_instances_ = task->GetCounter("instances");
+    num_transfers_ = task->GetCounter("alias_transfers");
+    num_zero_transfers_ = task->GetCounter("alias_zero_transfers");
+    num_instance_transfers_ = task->GetCounter("alias_instance_transfers");
+  }
+
+  void Process(Slice key, const Frame &frame) override {
+    MutexLock lock(&mu_);
+
+    // Get index for entity.
+    int index;
+    string id(key.data(), key.size());
+    auto fe = entity_mapping_.find(id);
+    if (fe == entity_mapping_.end()) {
+      index = entity_table_.size();
+      entity_table_.emplace_back(id);
+      num_entities_->Increment();
+      entity_mapping_[id] = index;
+    } else {
+      index = fe->second;
+    }
+
+    // Add aliases.
+    for (const Slot &s : frame) {
+      if (s.name == n_alias_) {
+        // Check language.
+        Frame alias(frame.store(), s.value);
+        if (alias.GetHandle(n_lang_) != language_) continue;
+        num_aliases_->Increment();
+
+        // Compute phrase fingerprint.
+        Text name = alias.GetText(n_name_);
+        int count = alias.GetInt(n_count_, 1);
+        int sources = alias.GetInt(n_sources_, 0);
+        int form = alias.GetInt(n_form_, 0);
+        uint64 fp = tokenizer_.Fingerprint(name);
+        if (fp == 1) continue;
+
+        // Look up or add phrase for entity to phrase table.
+        Phrase *&phrase = phrase_table_[fp];
+        if (phrase == nullptr) {
+          phrase = new Phrase(fp);
+          num_phrases_->Increment();
+        }
+
+        // Add entity to phrase.
+        bool reliable = (sources & reliable_alias_sources_);
+        phrase->entities.emplace_back(index, count, form, reliable);
+
+        // Add alias count to entity frequency.
+        entity_table_[index].count += count;
+        num_instances_->Increment(count);
+      }
+    }
+  }
+
+  void Flush(task::Task *task) override {
+    // Prune phrase table by transfering unreliable aliases to reliable
+    // aliases for related items.
+    if (transfer_aliases_) {
+      LOG(INFO) << "Transfer aliases";
```

Comment:
Do you still need the LOG(INFO) statements in this method?

Reply:
I use it for verifying that alias transfer is enabled and it is also useful for timing the alias transfer.